Spigot-API-Patches/0305-per-viewer-chat-rendering.patch

```diff
@@ -0,0 +1,381 @@
++        super(player, async);
++        this.viewers = viewers;
++        this.recipients = new HashSet<>(Bukkit.getOnlinePlayers());
++        this.renderer = renderer;
++        this.message = message;
++    }
++
++    @Deprecated
++    AbstractChatEvent(final boolean async, final @NotNull Player player, final @NotNull Set<Player> recipients, final @NotNull Set<Audience> viewers, final @NotNull ChatRenderer renderer, final @NotNull Component message) {
++        super(player, async);
++        this.recipients = recipients;
++        this.viewers = viewers;
++        this.renderer = renderer;
++        this.message = message;
++    }
++
++    @Deprecated
+     AbstractChatEvent(final boolean async, final @NotNull Player player, final @NotNull Set<Player> recipients, final @NotNull ChatComposer composer, final @NotNull Component message) {
+         super(player, async);
+         this.recipients = recipients;
++        final Set<Audience> audiences = new HashSet<>(recipients);
++        audiences.add(Bukkit.getConsoleSender());
++        this.viewers = audiences;
+         this.composer = composer;
+         this.message = message;
+     }
+@@ -33,10 +62,31 @@ public abstract class AbstractChatEvent extends PlayerEvent implements Cancellab
+     AbstractChatEvent(final boolean async, final @NotNull Player player, final @NotNull Set<Player> recipients, final @NotNull ChatFormatter formatter, final @NotNull Component message) {
+         super(player, async);
+         this.recipients = recipients;
++        final Set<Audience> audiences = new HashSet<>(recipients);
++        audiences.add(Bukkit.getConsoleSender());
++        this.viewers = audiences;
+         this.formatter = formatter;
+         this.message = message;
+     }
+ 
++    /**
++     * Gets a set of recipients that this chat message will be displayed to.
++     *
++     * <p>The set returned is not guaranteed to be mutable and may auto-populate
++     * on access. Any listener accessing the returned set should be aware that
++     * it may reduce performance for a lazy set implementation.</p>
++     *
++     * <p>Listeners should be aware that modifying the list may throw {@link
++     * UnsupportedOperationException} if the event caller provides an
++     * unmodifiable set.</p>
++     *
++     * @return a set of {@link Audience audiences} who will receive the chat message
++     */
++    @NotNull
++    public final Set<Audience> viewers() {
++        return this.viewers;
++    }
++
+     /**
+      * Gets a set of recipients that this chat message will be displayed to.
+      *
+@@ -49,22 +99,60 @@ public abstract class AbstractChatEvent extends PlayerEvent implements Cancellab
+      * unmodifiable set.</p>
+      *
+      * @return a set of players who will receive the chat message
++     * @deprecated in favor of {@link #viewers()}
```

Comment:
Add a note that *ALL* deprecated members in these classes will be removed with 1.17.